src/field/vecops.cxx

```diff
@@ -120,28 +160,31 @@ const Field2D Div(const Vector2D &v, CELL_LOC outloc) {
   result += DDY(metric->J*vcn.y, outloc);
   result += DDZ(metric->J*vcn.z, outloc);
   result /= metric->J;
-  
+
   return result;
 }
 
 const Field3D Div(const Vector3D &v, CELL_LOC outloc) {
   TRACE("Div( Vector3D )");
 
+  if (outloc == CELL_DEFAULT) {
+    outloc = v.getLocation();
+  }
+
+  ASSERT1(outloc != CELL_VSHIFT);
+
   Mesh *localmesh = v.x.getMesh();
   Field3D result(localmesh);
 
   Coordinates *metric = localmesh->coordinates(outloc);
 
-  if(outloc == CELL_DEFAULT) 
-    outloc = CELL_CENTRE;
-
   // get contravariant components of v
   Vector3D vcn = v;
   vcn.toContravariant();
-  
-  result = DDX(metric->J*vcn.x, outloc);
-  result += DDY(metric->J*vcn.y, outloc);
-  result += DDZ(metric->J*vcn.z, outloc);
+
+  result = DDX(metric->J * vcn.x, outloc);
+  result += DDY(metric->J * vcn.y, outloc);
+  result += DDZ(metric->J * vcn.z, outloc);
   result /= metric->J;
```

Comment:
We need to set the location of J here (note we do allow `v.getLocation==CELL_VSHIFT`). I think we need:
```
  result = DDX(vcn.x.getCoordinates()->J * vcn.x, outloc);
  result += DDY(vcn.y.getCoordinates()->J * vcn.y, outloc);
  result += DDZ(vcn.z.getCoordinates()->J * vcn.z, outloc);
```

Reply:
I am coming around to thinking that if `outloc` is `CELL_DEFAULT` and `v` is at `CELL_VSHIFT`, then the result location should be `CELL_CENTRE`, as that's the only location that we currently support, at least for scalar-result operations.

On the other hand, for vector-result operations, you might expect the result to be at the same location as the input, rather than `CENTRE`, which is not possible for `VSHIFT` and in that case it is definitely better to fail.

Is it better to be consistent across vector operations, and require the user to be explicit? Or better to do "the right thing"?

Reply:
@ZedThree I don't have a strong opinion: I suspect there is nobody using CELL_VSHIFT anyway. For what it's worth, my first thought was that `Div` on a CELL_VSHIFT vector would default to a result at CELL_CENTRE...